# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [4]:
# Importar pandas
import pandas as pd

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, Ricardo, muy buen trabajo importando ``pandas``.</div>

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [5]:
# Leer el archivo y almacenarlo en df
df=pd.read_csv('/datasets/music_project_en.csv')

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo cargando los datos necesarios para el desarrollo del proyecto.⁣</div>

Muestra las 10 primeras filas de la tabla:

In [6]:
# Obtener las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con el ``head()``, por otro lado, ten en cuenta que en notebooks de Jupyter se podría usar ``display()`` en vez de ``print()`` a la hora de mostrar las tablas para generar una salida mucho más estética.⁣</div>


Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [7]:
# Obtener la información general sobre nuestros datos
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, muy buen trabajo con el uso del método ``info()``.</div>

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `En el caso de userID es mejor separarlo por medio de guion bajo dejandolo como :user_id`.




### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? 
      ¿Y cómo podemos entender lo que almacenan las columnas?`
               Se tiene por usuario la informacion general de la canción que se reproduce, a la hora exacta, en la cuidad y el        dia, con esto se puede entender algún patrón de actividad. 

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`
      La ciudad, la hora , y el dia de la semana son datos importantes para nuestra hipotesis

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`hay valores ausentes. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con las respuestas, Ricardo. Tal vez como observación  adicional podrías tener en cuenta que en la primera pregunta sería interesante determinar qué tipo de datos tienen las columnas de la tabla y si son el ideal o no. </div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [8]:
# Muestra los nombres de las columnas
print(df.columns)


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Bien hecho, buen trabajo mostrando cómo se ve cada uno de los encabezados de la tabla.</div>

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [9]:
# Bucle en los encabezados poniendo todo en minúsculas
new_columns_names=[]         #Se declara una lista vacia 
for old_name in df.columns:   #Se inicia el bucle en las columnas del DataFrame
    name_lowered=old_name.lower()   #Se utiliza el comando lower() para cambiar a minusculas
    new_columns_names.append(name_lowered)     #se añade los nombres a minusculas a la lista vacia
    
df.columns=new_columns_names                  #y se reasigna al DataFrame
print(df.columns)                             #Se muestran los cambios 


Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo con el bucle usado para aplicar el lowering a cada encabezado.</div>

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [10]:
# Bucle en los encabezados eliminando los espacios
new_columns_names=[]                      #Se declara una lista vacia 
for old_name in df.columns:                #Se inicia el bucle en las columnas del DataFrame
    name_strip=old_name.strip()            #Se utiliza el comando strip() para eliminar espacios en blanco 
    new_columns_names.append(name_strip)    #se añade los nombres modificados a la lista vacia
    
df.columns=new_columns_names                # se reasigna al DataFrame
print(df.columns)                           #Se muestran los cambios


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo con el bucle usado para aplicar el strip a cada encabezado.</div>

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [11]:
# Cambiar el nombre de la columna "userid"
column_new={"userid":"user_id"}              #se declara el cambio en column_new
df.rename(columns=column_new,inplace=True)   #se utiliza el metodo rename() junto con inplace para que se hagan los cambios en una sola linea
print(df.columns)                            #se muestra en pantalla 

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo aplicando el ``snake_case`` a la varible ``user_id``.</div>

Comprueba el resultado. Muestra los encabezados una vez más:

In [12]:
# Comprobar el resultado: la lista de encabezados
print(df.columns)


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, Ricardo, muy buen trabajo solucionando todos los inconvenientes de los encabezados.  </div>

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [13]:
# Calcular el número de valores ausentes
print(df.isna().sum())                     #se utiliza el metodo isna() en conjunto con sum() para verificar valores ausentes 


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo mostrando la cantidad de valores nulos de cada una de las columnas.⁣</div>

No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [14]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track','artist','genre']                     #se hace la lista de las columnas a iterar para hacer cambios 

for col in columns_to_replace:                                      #se hace la iteracion sobre las columnas y por el metodo replace se reemplazan los valores ausentes por 'unknown'
    df[col].fillna('unknown', inplace=True) 


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [15]:
# Contar valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo solucionando el problema de los valores nulos dentro de la tabla.⁣</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [16]:
# Contar duplicados explícitos
df.duplicated().sum()                #se verifica con el metodo duplicated() en conjunto con sum() para saber el numero de duplicados explícitos 


3826

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo identificando la cantidad de registros duplicados dentro de la tabla.</div>

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [17]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)    #con el metodo drop_duplicates() podemos eliminar los duplicados explícitos

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    

Muy bien Ricardo, recuerda que una vez se eliminan duplicados explícitos, lo más aconsejable es que debes reindexar la tabla para no tener problemas con esto en un futuro, lo puedes hacer así:
```python
df = df.drop_duplicates().reset_index(drop=True)
```
</div>

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [18]:
# Comprobar de nuevo si hay duplicados

print(df.duplicated().sum())

0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [21]:
# Inspeccionar los nombres de géneros únicos
print(df['genre'].sort_values().unique())                   #Esto se puede realizar con el método unique() que nos devuelve valores unicos en la columna seleccionada 
                                                            #Se agrega el metodo.sort_values()

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    

~~Buen trabajo, Ricardo; sin embargo, recuerda que el ideal es que los valores únicos de la columna se vean de manera ordenada alfabéticamente.~~
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    

Perfecto, buen trabajo Ricardo.
</div>

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [159]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genres):            # pasa una lista de valores incorrectos (wrong_genres),con el valor correcto (correct_genres) en la entrada de la función
    for wrong_g in wrong_genres:
        df['genre']=df['genre'].replace(wrong_g, correct_genres)  #se utiliza replace para hacer la corrección de los valores dados en la función     
    return df['genre']



Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [ ]:
# Eliminar duplicados implícitos
wrong=['hip','hop','hip-hop']                      #valores erroneos 
good='hiphop'                                      #valor correcto 
df['genre']=replace_wrong_genres(wrong,good)       #llamado de funcion previamente definida 
print(df['genre'])                                 #Se muestra en pantalla el resultado 



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    

Buen trabajo aplicando la función para solucionar el problema de los valores duplicados implícitos dentro de la columna.
</div>

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [161]:
# Comprobación de duplicados implícitos
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb',
       'hiphop', 'jazz', 'postrock', 'latin', 'classical', 'metal',
       'reggae', 'triphop', 'blues', 'instrumental', 'rusrock', 'dnb',
       'türk', 'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~Nuevamente, recuerda que el ideal es que los valores únicos de la columna se vean de manera ordenada alfabéticamente.~~
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    

Buen trabajo Ricardo.
</div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

Hay herramientas como duplicated que te ayudan a procesa datos duplicados y sobre la eliminacion de datos es muy util drop_duplicates(), en el caso de palabra compuestas( como hip hop) pueden quedar palabras sueltas; ahi es donde se implementan las funciones para procesar ese tipo de datos,al final pudieron tener resultados mas finos que al inicio. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    

Buen trabajo con las observaciones generadas Ricardo. </div>

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [162]:
# Contar las canciones reproducidas en cada ciudad
print(df.groupby(by='city')['track'].count())       #con el metodo groupby() en conjunto con by= podemos segmentar los datos para hacer el conteo

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`Comenta tus observaciones aquí`
En Springfield hay 42741 canciones reproducidas, mucho mas significativo que en Shelbyville

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [163]:
# Calcular las canciones reproducidas en cada uno de los tres días
print(df.groupby(by='day')['track'].count())    

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`Comenta tus observaciones aquí` Los dias en lo que mas se reproducen son los dias Viernes con 21840, seguido de el día Lunes con 21354 y al final el dia Miercoles con 18059, podemos apreciar los dias de la semana que mas pueden infuir con la actividad  para escuchar música. 

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [164]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day,city):
    

    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    day_1=df[df['day']== day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    city_1=day_1[day_1['city']==city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    conteo=city_1['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return conteo

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [165]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday','Springfield')

15740

In [166]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday','Shelbyville')

5614

In [167]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks('Wednesday','Springfield')

11056

In [168]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks('Wednesday','Shelbyville')

7003

In [169]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday','Springfield')

15945

In [170]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday','Shelbyville')

5895

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

La hipotesis es correcta ya que la ciudad de Shelbyville tiene un patrón diferente al de Springfield , el cual el día con mas escuchas son los dias miercoles en el caso de Shelbyville y en el caso de Srpingfield es el día de menos escucha. 

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

`Resume aquí tus conclusiones sobre la hipótesis.`
Podemos concluir que dependiendo de la cuidad , se tiene diferente tipo de actividad, ya que los numeros son muy diferentes entre cuidades,así mismo el día de la semana es muy importante para definir tambien las diferencias de acrividades entre Shelbyville y Springfield, ya que en Springfield el mayor numero de escuchas se encuentra el día Viernes dando el signo que hay actividades que permiten a la población escuchar mas musica ese día , de la misma forma en Shelbyville el mayor número de escuchas es el Miércoles, dando así la conclusión de que en efecto la actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad. 

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)